### ISSUE TITLE
API Server (and clients) becomes unresponsive with too many CRDs

### ISSUE THREAD

**Issue by ulucinar** [crossplane/crossplane#2649]
### What problem are you facing?

As part of the ongoing [Terrajet-based] providers effort, we have observed that registering 100s of CRDs has some performance impacts on the cluster. Observations from two sets of experiments are described [here](https://github.com/crossplane-contrib/terrajet/issues/47#issuecomment-920316141) and [here](https://github.com/crossplane-contrib/terrajet/issues/47#issuecomment-920441882). As discussed in the results of the first experiment, Kubernetes scalability thresholds document currently does not consider #-of-CRDs (per cluster) as a dimension. However, `sig-api-machinery` folks suggest a maximum limit of 500 CRDs. And the reason for this suggested limit is not API call latency SLOs but rather, as we also identified in our experiments, due to the latency in the OpenAPI spec publishing. As the results of the second experiment demonstrate, the marginal cost of adding a new CRD increases as more and more CRDs exist in the cluster.

Although not considered as a scalability dimension officially yet, it looks like we need to be careful for the #-of-CRDs we install in a cluster. And with Terrajet-based providers we would like to be able to ship 100s of CRDs per provider package. Currently, for the initial releases of these providers, we are including a small subset (less than 10% of the total count) of all the managed resources we can generate. We would like to be able to ship all supported resources in a provider package.

[Terrajet-based]: https://github.com/crossplane-contrib/terrajet

### How could Crossplane help solve your problem?

As part of the `v1.Provider`, `v1.ProviderRevision`, `v1.Configuration` and `v1.ConfigurationRevision` specs, we could have a new configuration field that tells the GVKs of package objects to be installed onto the cluster. To behave backwards-compatible, if this new API is not used, all objects defined in the package manifest get installed. If the new field has a non-zero value, then it's enabled and only selected objects get installed. In order to make the UX around installing packages using the new configuration field easier, we can also add new options for the `install provider` and `install configuration` commands of the Crossplane `kubectl` plugin.

As described in the experiment results, when a large #-of-CRDs are registered in a relatively short period of time, the background task that prepares the OpenAPI spec to be published from the `/openapi/v2` endpoint may cause a spike in the API server's CPU utilization, and this may saturate the CPU resources allocated to the API server. Similar to what controller-runtime client does, we may also consider implementing a throttling mechanism in the package revision controller to prevent this. However, because of the reasons discussed above and in the experiment results (especially the latency introduced in OpenAPI spec publishing), it looks like we will need a complementary mechanism like the one suggested above in addition to a throttling implementation in the package revision reconciler.

**Comment by negz** [comment]
Breadcrumbs for this possibly related issue: https://github.com/kubernetes/kubernetes/issues/101755

**Comment by negz** [comment]
@muvaf I'm not convinced the second part of what you've mentioned - users have to install more CRDs than they need - is a real problem. If there were no performance downsides would it have any other meaningful impacts?

Presumably the main downside would be 'clutter' in OpenAPI discovery and `kubectl get crds`, but that seems like a very minor concern. Especially since if we do start having folks filter the set of installed CRDs those folks are presumably going to have to remember to go update their provider as they begin to need API groups that they didn't think they would before.

If we are indeed prioritising this mostly because it's a quick win for performance while we look into other options, could we do it without making a change to the `Provider` API? e.g. Add the flag to provider binaries and advise folks to use a `ControllerConfig` to set that flag if they're hitting performance issues, until we get long term fixes in place?

**Comment by negz** [comment]
It seems like in https://github.com/kubernetes/kubectl/issues/1126 there's consensus that the rate limiting in `kubectl` will be removed and they'll instead defer to [APF](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/). I believe this should alleviate the discovery rate limiting issue - we'd be subject to configurable server side request queueing instead.

I've raised https://github.com/kubernetes-sigs/controller-runtime/issues/1707 to discuss whether a similar change to controllers would make sense.

**Comment by negz** [comment]
### Executive summary:

* Rate limiting CRD installs doesn't reliably alleviate the problems we're seeing.
* Fixes are open for the problems we're seeing and _may_ be available as patch releases in ~1 month.
* I support breaking Terrajet into smaller providers rather than adding filtering support.

Sadly I'm feeling convinced that at the moment the only way to reliably avoid the API server and `kubectl` performance issues we're seeing - apart from waiting for the upstream fixes - is to avoid creating 'too many' CRDs. It seems like what constitutes too many varies depending on the resources available to the control plane so there's not a reliable number we can cap it at, but ~300 seems to be a conservative estimate at the moment.

### Rate Limiting Experiments

GKE continues to be unable to scale to 2,000 CRDs (regardless of how we rate limit) without becoming unresponsive. API discovery begins to suffer from ~20 second long client-side rate limiting with as few as 200 CRDs in the cluster. The most successful strategy I've found so far for GKE is batches of 50 CRDs spread 30 seconds apart.

EKS on the other hand will happily accept 2,000 CRDs applied all at once with no rate limiting at all or other immediately discernible performance degradation but will then exhibit client-side rate limiting of up to _six minutes_ before some kubectl commands will complete.

In my experience `kind` does seem to work (I got it up to 4,000 CRDs) but uses a huge amount of resources (2 cores, GBs of memory) which is not a great experience for someone wanting to try out Crossplane for the first time on their laptop or workstation.

I've noticed that CPU consumption seems to drop off about an hour after a huge number of CRDs are applied; presumably this is how long the API server takes to recompute its OpenAPI schema over and over again. Memory consumption doesn't seem to drop unless the API server process restarts. The API server has a special clause to load many CRDs more efficiently at startup as compared to those same CRDs being added at runtime.

### Impending Upstream Fixes

Both of the key upstream issues we're facing (OpenAPI processing and kubectl discovery rate limiting) have PRs open (https://github.com/kubernetes/kube-openapi/pull/251, https://github.com/kubernetes/kubernetes/pull/105520). I would expect these fixes to be merged imminently but there are no guarantees. If they're candidates for patch releases it's possible fixes will be available within a month per the [patch release schedule](https://kubernetes.io/releases/patch-releases/#cadence). I've reached out for clarification around whether the folks working on the issues expect them to be backported and available as patch releases, or whether they'll need to wait until the next minor release.

If the upstream fixes do indeed become available as patch releases I personally would feel comfortable requiring that Crossplane users be on a supported version of Kubernetes at the latest patch release in order to support large providers. Asking users to update to the latest minor version of Kubernetes seems like a taller order and not something many enterprises would be able to do easily.

### Options for Reducing Installed CRDs

Personally I still don't buy that there's any real value in reducing the number of CRDs (used or not) in the system _except_ to workaround these performance issues, but it seems like said performance issues alone may force our hand.

> have we considered just breaking these large providers into smaller, group-based providers? It would require no code changes, just some different flags in the package build process to only include some CRDs and maybe set some flag on the entrypoint of the controller image.

Of the two reduction avenues I'm aware of (smaller providers vs filtering what CRDs are enabled for a provider) I support the approach @hasheddan proposed above. It sounds like we'd need to work through a few things technically to make it work though, as @muvaf mentioned. For example:

* Do we keep one `ProviderConfig` per provider (now group scoped) or try to share across providers?
* What do we do with cross resource references, which would now mostly be to other providers (i.e. API groups)?

I feel the benefits of smaller providers (vs filtering) are:

* It requires no changes to upstream Crossplane, decoupling it from the Crossplane review and release cycle.
* Given the goal is to reduce the ratio of installed to actively used CRDs it seems (IMO) much more intuitive to "install the providers for the functionality you need". The alternative feels more to me like requiring users to carefully and actively avoid installing too many CRDs.
* It's not something we need to commit to; for example we could return to larger providers (if so desired) or simply not worry about providers growing larger once the upstream issues were fixed.

Of course, neither smaller providers nor filtering providers will actually fix the scalability issues - they'll just reduce the likelihood that folks will run into them so ultimately we are going to need to continue working with upstream Kubernetes to ensure the API server can meet our needs.

**Comment by negz** [comment]
Turns out at the time of writing the proposed kubectl fix (https://github.com/kubernetes/kubernetes/pull/105520) doesn't work, but I raised https://github.com/kubernetes/kubernetes/pull/106016 which does. Discovery time is reduced from 6 minutes to 15 seconds in my tests with ~2.000 CRDs.

**Comment by muvaf** [comment]
Findings are from before @negz posted and they are pretty similar, so you can skip to suggestions section.

---

### Findings

I ran some tests in `kind` cluster where the CPU and memory have hard limits to see the behavior. [This Grafana dashboard](https://snapshot.raintank.io/dashboard/snapshot/HDamUxs2qW3LAZjeDMTgwFydfgbilK6D) shows what happens when you create 750 CRDs with `1s` wait period. A summarizing screenshot where the creation of CRDs starts at 18:03 and finishes at 18:16:
![image](https://user-images.githubusercontent.com/7584126/139422489-5df76feb-d542-425a-8cab-8309d8082d15.png)

As you can see even thought the creation calls finish in ~13m, the burst takes around ~45-50m. In addition, we see that even after the burst we see increased memory consumption (from `450 MB` to `3 GB`), relatively insignificant CPU consumption and big increase in the number of Go routines (from ~3K to ~18K). Another interesting point that is not immediately clear from the dashboard is that during the burst;
* Error budget decreases,
* Write availability takes a hit (from 100% to ~99,998%),
* Read availability takes a rather smaller hit.
* `crd_openapi_controller`, `APIServiceRegistrationController`  and `open_api_aggregation_controller`  reach 10s latency and stay there for a long while.

I have done many experiments, including applying CRDs with empty schema, but the behavior is roughly the same. @negz and @ulucinar have dove [into the code](https://github.com/kubernetes/kubernetes/blob/0fb71846df9babb6012a7fce22e2533e9d795baa/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go#L44) of these controllers and the main suspect is that each reconciliation tries to re-calculate cumulative OpenAPI v2 schema independently and at the same time. The more overlap these processings have, the higher the burst is.

To summarize my findings:
* The burst happens because of overlaps of the processings.
* During the burst, `api-server` is either very slow or unresponsive at all.
* After the burst, we see very high increase in memory usage depending on the number of CRDs added.
* The calculations are related to how big of the schema CRD has but even empty schema causes similar behavior.

Additionally, as @negz commented above, some API servers are able to take the load, like EKS clusters or regional GKE clusters to some extent.

---

### Suggestions

I think @negz also reached a similar conclusion with his tests in GKE and EKS; IMO, while it's able to take the load in some cases, it is just too easy to break the cluster even with different variations of throttling and there is no condition on the CRD that we can use to track whether we should continue applying new CRDs. In addition, the additional CRDs impose continuous memory load on the api-server even after the installation finishes and the burst ends.

> There is one idea that I'd like to test over the weekend though: apply 100 CRDs, then in each reconcile pull `/openapi/v2` JSON and continue with the next 100 only if all that was applied in the former batch shows up in the JSON, which could mean the schema aggregation for that 100 is completed.

The following is roughly how I think each selective API installation approaches would work. Though I want to note that I think we'd apply those options only to the providers with more than ~250 CRDs because both of them causes degradation in UX to some extent.

* Different packages for each API group.
    * We’d have a `provider-tf-aws-base` that has required and most referenced groups, like `ProviderConfig`, `EC2`, `KMS` etc. and not include those in others.
    * Users need to make sure dependent provider versions are compatible schema-wise so that any sort of referencing (including cross-resource, (if implemented) generic resource references and indirect composition patches) work when you update provider A while referencing its resources from resources of provider B.
    * Also some providers will drop CRDs, especially in v1alpha ones, so integrity of that dependency is important.
    * We can just instruct users to do it initially but ideal case is to make providers declare dependencies like `Configuration`s do, i.e. `provider-tf-aws-eks:v0.20.0` depends on `>=provider-tf-aws-base:v0.20.0`
    * The number of API groups is not low. Provider-tf-aws has total of 141 API groups, meaning at least 120 `Provider` packages if we have a `provider-tf-aws-base`.
* Additional API groups on top of default ones.
    * Developers can add the default API groups to [`Provider.meta.pkg.crossplane.io`](https://github.com/crossplane/provider-aws/blob/12216a5/package/crossplane.yaml#L139) file and those groups would always be installed by the package manager.
        * If there is no `apiGroups` list, `*` will be assumed.
    * For additional ones, users can list them in `Provider` CR.
    * The format would be almost same as [how Kubernetes RBAC selects API groups](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-examples), except the `verbs`.
    * The default `apiGroups` in meta file and the `apiGroups` in `Provider` resource would be aggregated and the final list of CRDs would be installed.
    * `Configuration` [dependencies](https://github.com/upbound/platform-ref-aws/blob/764720c/crossplane.yaml#L92) will have that `apiGroups` as well to make sure `Provider` is installed with the `apiGroups` that `Configuration` uses.

As pointed out earlier, there is a PR that might fix the problem and I know @ulucinar is planning on validating that. I think the result of his tests will be one of the key factors for our decision here.

If the fix works for us and drops as a patch release to supported k8s versions, I think there is a consensus that we can let users know that big providers can be installed reliably only with the latest patched versions and not implement any selective API approaches today. I think we'll need it at some point because we have runtime costs as well but that can wait until users hit that problem (my RPi 4 cluster will hit these problems immediately but I'm fine, don't worry).

If the fix works but drops as a minor release, meaning users can’t really deploy those big providers in production for several months, then I believe we should go with package separation only TF providers but in a less granular way that has the least reference possibility, i.e. in total of 3 packages for `141` groups `provider-tf-aws` has instead of `>100` packages, with the assumption that once the fixed versions becomes default, we can merge them back. Though that’s some painful surgery.

If the fix doesn’t work though, I’m feeling much closer to the additional API groups approach. I think package separation won’t work without a dependency mechanism and I can’t really justify that additional management complexity in my mind with more than 100 API groups for a single big provider like AWS. When I think of OS package managers, this dependency is managed automatically but I’d not feel comfortable that updating provider A should cause update of provider B automatically in Crossplane.

**Comment by muvaf** [comment]
> There is one idea that I'd like to test over the weekend though: apply 100 CRDs, then in each reconcile pull `/openapi/v2` JSON and continue with the next 100 only if all that was applied in the former batch shows up in the JSON, which could mean the schema aggregation for that 100 is completed.

I have [tested](https://github.com/crossplane-contrib/provider-tf-aws/pull/127/commits/5e907597b584a41987eb093c253a69e378c2f4f3#diff-2960dea708e16d16821e2cbd47f3be21db8dcd4a9e313b01d9c01da3d1edce98) this idea. It downloads and checks for a GVK string existence in the JSON blob and that reliably reports whether all processing has been done for given CRD. You can follow the instructions [here](https://github.com/crossplane-contrib/provider-tf-aws/pull/127/commits/5e907597b584a41987eb093c253a69e378c2f4f3#diff-38a9d02c990a7f1bfc5e8e772d175d9d3e4b6a857421943825bc71e565cac11c) to repeat it with different variables, like batch size, wait period between batches, cooldown period etc.

With that information, **we are able to tame the burst via throttling**. After each batch, it doesn't move to the next one until it can confirm all CRDs exist in the OpenAPI JSON and if they do, we see CPU and memory dropped to normal levels.

However, my suspicion about the possibility that the calculations are repeated because all reconcile work in parallel to update the same JSON, hence taking longer is proven wrong. Tested with different batches in both `kind` and GKE zonal cluster, but, overall **the total duration did not change significantly** even if we were able to throttle properly with a reliable handle.

---

There are 4 different results in the [draft PR](https://github.com/crossplane-contrib/provider-tf-aws/pull/127):
* `kind` batch size 50, `10s * batch no` wait period between batches.
* `kind` batch size 100, no wait between batches.
* `GKE Zonal` batch size `1`, no wait between batches.
* `GKE Zonal` batch size `50`, `5s * batch no` wait period between batches.

GKE ones never finished because around 150-200 CRDs, api-server times out when we request the openapi JSON.

The most notable one is the following and we see similar pattern in all tests in general. In all other tests, we see CPU banging `3` limit but here we are able to make it not hit that limit:


![image](https://user-images.githubusercontent.com/7584126/139598654-bfefb3b4-95f1-4f38-90cd-de185c073dbc.png)

As you can see, we're able to smooth it out, but in the end the total duration doesn't really change:
```
started at 2021-10-30 23:18:39.761856 +0300 +03 m=+1.579766166
let me check if 50 CRDs made it to OpenAPI schema
could not find 0, waiting for 10s before retrying...
could not find 23, waiting for 10s before retrying...
could not find 45, waiting for 10s before retrying...
the batch with 50 CRDs took 30.719723984s to install


 51 out of 752 CRDs have been installed. Waiting for 10s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 40.944071864s to install


 101 out of 752 CRDs have been installed. Waiting for 20s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 1m1.674940509s to install


 151 out of 752 CRDs have been installed. Waiting for 30s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 1m11.657697653s to install


 201 out of 752 CRDs have been installed. Waiting for 40s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 1m32.686731552s to install


 251 out of 752 CRDs have been installed. Waiting for 50s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 1m42.520113184s to install


 301 out of 752 CRDs have been installed. Waiting for 1m0s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 2m14.331273956s to install


 351 out of 752 CRDs have been installed. Waiting for 1m10s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 2m34.84863816s to install


 401 out of 752 CRDs have been installed. Waiting for 1m20s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 2m47.421192317s to install


 451 out of 752 CRDs have been installed. Waiting for 1m30s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 3m8.740966523s to install


 501 out of 752 CRDs have been installed. Waiting for 1m40s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 3m28.447167459s to install


 551 out of 752 CRDs have been installed. Waiting for 1m50s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 3m30.408228467s to install


 601 out of 752 CRDs have been installed. Waiting for 2m0s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 3m40.051317962s to install


 651 out of 752 CRDs have been installed. Waiting for 2m10s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 3m51.713754958s to install


 701 out of 752 CRDs have been installed. Waiting for 2m20s to start the next batch...

let me check if 50 CRDs made it to OpenAPI schema
...
the batch with 50 CRDs took 5m10.730401149s to install


 751 out of 752 CRDs have been installed. Waiting for 2m30s to start the next batch...

Checking schema for all CRDs...
could not find 750, waiting for 10s before retrying...
could not find 750, waiting for 10s before retrying...
the batch with 752 CRDs took 35.872272471s to install
finished at 2021-10-31 00:16:38.707375 +0300 +03 m=+3480.476232410
the whole installation took 57m58.896466244s

```

**Comment by ulucinar** [comment]
Here are some initial results from two sets of experiments performed on kind clusters running on a GCP `e2-standard-4` VM. 

The first experiment does **not** include the [lazy marshaling PR](https://github.com/kubernetes/kube-openapi/pull/251) under consideration and is included here as a reference. Similar to our previous results reported [here](https://github.com/crossplane-contrib/terrajet/issues/47#issuecomment-920316141) and in the comments above, we have the expected high CPU/memory utilization period following the provisioning of the 658 CRDs from `provider-tf-azure` with no client-side throttling. The kind node image used for this experiment is `ulucinar/node-amd64:no-lazy-marshaling`:

<img width="980" alt="no-lazy-marshal-metrics" src="https://user-images.githubusercontent.com/9376684/139755501-327fd474-0d7d-40a1-954c-001876c07a9a.png">

The second experiment has the lazy marshaling PR incorporated and is run on the same VM. The kind node image built for this experiment is `ulucinar/node-amd64:lazy-marshaling`. We can clearly observe improvements in both CPU and memory usage here:

<img width="979" alt="lazy-marshal-metrics" src="https://user-images.githubusercontent.com/9376684/139756613-2f1d3596-b523-4039-b123-7d86f5e0e909.png">

**Comment by apelisse** [comment]
That's really awesome. Quick question: there is a massive difference in the time it took for the CRD count to increase, is it because the CRDs applied much faster or is it something else? Is it a cause or a consequence? I'm not sure how to interpret this data.

**Comment by ulucinar** [comment]
Hi @apelisse,
Thank you very much. In both experiments, a `kubectl create -f` is run on a directory containing `658` CRDs, with no "external" client-side throttling, excluding of course any client-side throttling that may be performed by the controller-runtime client itself. Perhaps the legend & the title in "CRD Count" graphs are a bit misleading. These are the graphs of the [`apiextensions_openapi_v2_regeneration_count` counter](https://github.com/kubernetes/kubernetes/blob/a2960dc2775848f3b54fb288650e02b4f691743b/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go#L193) with `reason == "add"` vs. time. My interpretation is that with the [proposed PR](https://github.com/kubernetes/kube-openapi/pull/251), the [OpenAPI controller](https://github.com/kubernetes/kubernetes/blob/a2960dc2775848f3b54fb288650e02b4f691743b/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go#L43) can process its queue faster because of deferred marshaling.

**Comment by ulucinar** [comment]
We (@muvaf, @hasheddan and me) attended today’s sig-api-machinery meeting held at 11:00am PST to discuss whether it's possible to cherry-pick the [lazy marshaling PR](https://github.com/kubernetes/kube-openapi/pull/251) to active release branches.

So the general consensus is that the lazy marshaling PR is close to be merged after some remaining issues with the PR are addressed. But it looks like the PR will be merged soon with a high probability.

Regarding the cherry-picks to active release branches, the sig-api-machinery folks are positive about those cherry-picks but they also mentioned that after the PR is merged, the associated risks with backporting need to be evaluated. But they also mentioned that given the small size of the PR, and given that it only tweaks when the aggregated OpenAPI spec is computed (it’s now computed lazily when a request arrives), the risks should not be high.

**Comment by negz** [comment]
Some important dates for getting our proposed fixes released. We have two fixes in flight:

https://github.com/kubernetes/kubernetes/pull/106181 - We hope to have this cherry-picked so that our users aren't required to update their API servers to Kubernetes v1.23 in order to install many CRDs.

https://github.com/kubernetes/kubernetes/pull/105520 (or https://github.com/kubernetes/kubernetes/pull/106016) - This client-side (`kubectl`) fix won't be cherry-picked - we hope to have it merged in time to be included with Kubernetes v1.23. We feel that `kubectl` users should be able to update pretty easily.

The timelines we need to meet are:

* Friday, Nov 12 - November patch release cherry-pick deadline
* Friday, Nov 17 - November patch release cherry-picks are released
* Tuesday, Nov 16 - v1.23 code freeze (kubectl fix must be merged before this date)
* Tuesday, Dec 7 - v1.13 is released

Fallback options:

* Friday, Dec 10 - December patch release cherry-pick deadline
* Wed, Dec 15 - December patch releases are released.

Note that if we miss the deadline for the Nov 17th cherry picks we could fall back to the December

**Comment by negz** [comment]
@roycaihw just provided the following input on Kubernetes Slack, that may help us create more scoped cherry-pick PRs:

> Not sure if there is an easier way using go mod today. But here are some examples for how we used to do a more concrete cherrypick w/ kube-openapi (instead of bumping all of kube-openapi):
> * https://github.com/kubernetes/kube-openapi/pull/68
> * https://github.com/kubernetes/kubernetes/pull/63627
> Basically, to cherrypick https://github.com/kubernetes/kube-openapi/pull/251 to k/k 1.21:
> * Create a branch A in kube-openapi, tracking the kube-openapi version used in k/k 1.21
> * Cherrypick PR#251 to branch A
> * In k/k, create a PR to update 1.21 to use branch A

**Comment by negz** [comment]
The general consensus within the folks working on this (i.e. @ulucinar, @muvaf, @luebken, and myself) is that we're not likely to get these cherry-picked in time for the patch releases that will be released on Nov 17th:

* The PR that needs cherry-picking is not yet reviewed or merged (but is fairly uncontroversial).
* The deadline to get those cherry-picks merged is the 12th (this Friday).
* The 11th is a public holiday for many folks here in the US.
* We've been advised by @hasheddan (sig-release team emeritus) that the closer you get to the cherry-pick deadline, the harder a sell it is to cherry-pick a PR.

**Comment by ulucinar** [comment]
Just left a comment in the upstream PR to signal the direction we would like proceed with:
https://github.com/kubernetes/kubernetes/pull/106181#issuecomment-963527818

**Comment by negz** [comment]
Quoting @apelisse [on Kubernetes Slack](https://kubernetes.slack.com/archives/C0EG7JC6T/p1636405390158100?thread_ts=1634838218.016300&cid=C0EG7JC6T):

> If you can get this done quickly enough to get your cherry-picks in before cherry-pick freeze, I'll happily approve. I'm mostly focusing on the main release code-freeze though so we won't have time to push that ourselves

He's helped us by creating the following kube-openapi branches:

* https://github.com/kubernetes/kube-openapi/tree/release-1.20
* https://github.com/kubernetes/kube-openapi/tree/release-1.21
* https://github.com/kubernetes/kube-openapi/tree/release-1.22

These branches correspond to the kube-openapi commits Kubernetes 1.20, 1.21, and 1.22 are currently using. We will now proceed by:

1. Opening PRs to cherry-pick https://github.com/kubernetes/kube-openapi/pull/251 to each of the above branches.
2. Opening PRs to update the k/k release-1.20, release-1.21, and release-1.22 branches to the latest commit of their corresponding k/kube-openapi release branches.

This will reduce the scope of the k/k cherry-picks back down to only including https://github.com/kubernetes/kube-openapi/pull/251.

**Comment by muvaf** [comment]
The current status is that the scope of changes for release-1.20 and release-1.21 had included some dependency bump and we had to revert that. Now, they both have minimal scope just like release-1.22 one.

The PR that bumps master is merged, meaning the fix is **guaranteed to be included in 1.23**.

The patch release PRs are:
* 1.20: https://github.com/kubernetes/kubernetes/pull/106257
* 1.21: https://github.com/kubernetes/kubernetes/pull/106255
* 1.22 (approved for merge): https://github.com/kubernetes/kubernetes/pull/106250

Once we get approved-for-merge for 1.20 and 1.21 as well, we'll ping release manager to see if they can include them in the respective patch releases.

**Comment by muvaf** [comment]
All the patch release PRs are merged. Now it's guaranteed that the fix will be included in the following releases of Kubernetes:
* 1.23 (December 7th)
* 1.22.4 (November 17th)
* 1.21.7 (November 17th)
* 1.20.13 (November 17th)

I think we can close this issue once the kubectl PR https://github.com/kubernetes/kubernetes/pull/106016 is merged as well.

**Comment by ulucinar** [comment]
Here are kind node images built from the master & active release k/k branches containing the lazy marshaling behavior:
master: `ulucinar/node-amd64:master-lazy-marshal`
1.22: `ulucinar/node-amd64:1.22-lazy-marshal`
1.21: `ulucinar/node-amd64:1.21-lazy-marshal`
1.20: `ulucinar/node-amd64:1.20-lazy-marshal`

**Comment by ulucinar** [comment]
All branches now contain the OpenAPI aggregated spec lazy marshaling changes. Here are `kind` images for them:
```
ulucinar/node-amd64:v1.23.0-beta.0
ulucinar/node-amd64:v1.22.4
ulucinar/node-amd64:v1.21.7
ulucinar/node-amd64:v1.20.13
```

`kind` has also released an official node image for the `release-1.22` branch:
```
kindest/node:v1.22.4
```

**Comment by negz** [comment]
The sig-cli folks ended up merging https://github.com/kubernetes/kubernetes/pull/105520, which bumped up rate limits quite a lot, rather than disabling client-side rate limiting. Specifically:

* The discovery request rate was bumped from 50rps, up from 5.
* The discovery client now allows bursts of up to 300rps, up from 100.

**Comment by negz** [comment]
We may want to leave this issue open to track further upstream improvements; notably there's still an appetite to remove the client-side rate limiting (on discovery at least) in kubectl and also to improve how discovery is cached.

**Comment by muvaf** [comment]
@negz I think we should close this issue as the problems it's describing are fixed, i.e. clusters and clients do not become unresponsive with too many CRDs anymore. If there are other problems, a new issue that is more specific could make more sense.

**Comment by muvaf** [comment]
I'm closing this now. Feel free to re-open if you experience described problems in the released versions.

**Comment by jonnylangefeld** [comment]
@muvaf I just tested the newly merged client update with the increased burst and QPS that was referred by @negz, but I still ran into client side throttling. And that on a cluster with <300 CRDs.
All details on the original upstream PR: https://github.com/kubernetes/kubernetes/pull/105520#issuecomment-990582452

**Comment by muvaf** [comment]
@jonnylangefeld thanks for reporting! Either client side or server side throttling is expected during the cache calls but the part that we were interested in fixing was the client becoming unresponsive, i.e. 6mins of wait for a simple `GET` call or timing out. How long does it take for calls to complete in your case?

**Comment by jonnylangefeld** [comment]
A simple `GET` request via raw API call wasn't the issue before though right? That was always fast. Only if you do a `kubectl get`,  then the discovery cache ran and that is what took long. Here a comparison of the two:
```
time kubectl get pods -n default
kubectl get pods -n default  0.50s user 1.19s system 11% cpu 14.174 total

time kubectl get --raw  "/api/v1/namespaces/default/pods"
kubectl get --raw "/api/v1/namespaces/default/pods"  0.06s user 0.03s system 23% cpu 0.398 total
```

In the [other post](https://github.com/kubernetes/kubernetes/pull/105520#issuecomment-990582452) I describe how `kubectl get` does 170 requests to the API server if the doesn't cache exists and 4 requests if it does exist (`time kubectl get --raw  "/api/v1/namespaces/default/pods"` is only 1 request).

**Comment by apelisse** [comment]
There are improvements related to this that I'd love to see done @seans3 @justinsb

**Comment by jonnylangefeld** [comment]
I did a bit of digging and found out why the original upstream PR https://github.com/kubernetes/kubernetes/pull/105520 didn't work: https://github.com/kubernetes/kubernetes/pull/105520#issuecomment-997576747.

Maybe give my fix a thumbs up to make `kubectl` more usable again when controllers like crossplane are installed?

It's only one step on the way to at least not always run into rate limits. To actually optionally disable the discovery cache I opened a separate issue: https://github.com/kubernetes/kubernetes/issues/107130.

**Comment by lippertmarkus** [comment]
This is a real blocker for us. We're on 1.22.4 so applying the ~650 CRDs of provider-jet-azure is quite fast with the lazy marshaling change. On the client-side however, it's not just kubectl which is quite unusable with the throttling but also other clients like management dashboards (we're using Rancher) which get extremely slow.

Can we maybe revisit the idea of reducing installed CRDs?

**Comment by jonnylangefeld** [comment]
The fix I posted in the comment below will make it into kubectl 1.24, so that’ll be a slight improvement for kubectl. Any other client is still affected. Basically anything using client-go.
Check out this blog post for a in-depth explanation of the issue: https://jonnylangefeld.com/blog/the-kubernetes-discovery-cache-blessing-and-curse.

Unfortunately a real solution can only be achieved through a server change as well.
But if we want to keep building controllers we’ll have to fix this issue.

**Comment by apelisse** [comment]
Hey there,

Very aware of lots of problems with discovery, the large amount of requests and how the cache is not invalidated properly.
@jefftree and I have been designing a possible solution but we have limited cycles for implementation. Happy to help someone implement some changes if you have some bandwidth available. Also happy to discuss the problem in more depth to make sure we address it properly.

Thanks

**Comment by ulucinar** [comment]
Hi @apelisse,
I'm also investigating these issues and their implications. Would love to share ideas, gain a deeper understanding and contribute to the implementations. Thank you for your support!

**Comment by negz** [comment]
@apelisse We're happy to help if we can. Is there somewhere we can find your thoughts so far?
